src/NUnitTestAdapter/TestConverter.cs

```diff
@@ -40,61 +40,66 @@ public TestConverter(TestLogger logger, string sourceAssembly)
         /// using the best method available according to the exact
         /// type passed and caching results for efficiency.
         /// </summary>
-        public TestCase ConvertTestCase(ITest test)
+        public TestCase ConvertTestCase(XmlNode testNode)
         {
-            if (test.IsSuite)
+            if (testNode.Name != "test-case")
                 throw new ArgumentException("The argument must be a test case", "test");
 
             // Return cached value if we have one
-            if (vsTestCaseMap.ContainsKey(test.TestName.UniqueName))
-                return vsTestCaseMap[test.TestName.UniqueName];
+            string id = testNode.GetAttribute("id");
+            if (vsTestCaseMap.ContainsKey(id))
+                return vsTestCaseMap[id];
            
             // Convert to VS TestCase and cache the result
-            var testCase = MakeTestCaseFromNUnitTest(test);
-            vsTestCaseMap.Add(test.TestName.UniqueName, testCase);
+            var testCase = MakeTestCaseFromXmlNode(testNode);
+            vsTestCaseMap.Add(id, testCase);
             return testCase;             
         }
 
-        public TestCase GetCachedTestCase(string key)
+        public TestCase GetCachedTestCase(string id)
         {
-            if (vsTestCaseMap.ContainsKey(key))
-                return vsTestCaseMap[key];
+            if (vsTestCaseMap.ContainsKey(id))
+                return vsTestCaseMap[id];
 
-            logger.SendErrorMessage("Test " + key + " not found in cache");
+            logger.SendErrorMessage("Test " + id + " not found in cache");
             return null;
         }
 
-        public VSTestResult ConvertTestResult(NUnitTestResult result)
+        public VSTestResult ConvertTestResult(XmlNode resultNode)
         {
-            TestCase ourCase = GetCachedTestCase(result.Test.TestName.UniqueName);
+            TestCase ourCase = GetCachedTestCase(resultNode.GetAttribute("id"));
             if (ourCase == null) return null;
 
             VSTestResult ourResult = new VSTestResult(ourCase)
-                {
-                    DisplayName = ourCase.DisplayName,
-                    Outcome = ResultStateToTestOutcome(result.ResultState),
-                    Duration = TimeSpan.FromSeconds(result.Time)
-                };
+            {
+                DisplayName = ourCase.DisplayName,
+                Outcome = GetTestOutcome(resultNode),
+                Duration = TimeSpan.FromSeconds(resultNode.GetAttribute("duration", 0.0))
+            };
+
+            var startTime = resultNode.GetAttribute("start-time");
+            if (startTime != null)
+                ourResult.StartTime = DateTimeOffset.Parse(startTime);
+
+            var endTime = resultNode.GetAttribute("end-time");
+            if (endTime != null)
+                ourResult.EndTime = DateTimeOffset.Parse(endTime);
 
             // TODO: Remove this when NUnit provides a better duration
```

Comment:
Good idea. The resolution of our timers has increased for NUnit 3.0, so I was hoping that it would be possible to remove this code. It turns out to be better than it was but there are still some tests that show up with a zero duration. That makes the adapter list them as "Not Run" when the display is by duration.

I'll create an NUnit issue for this.